artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java

```diff
@@ -135,10 +130,16 @@ public void shutdown(long timeout, TimeUnit unit) {
          //can be set by just one caller.
          //As noted on the execute method there is a small chance that some tasks would be enqueued
       }
-      ArrayList<T> returnList = new ArrayList<>(tasks);
-      tasks.clear();
-
-      return returnList;
+      int pendingItems = 0;
+      //there is a small chance that execute() could race with this cleanup: the lock allow an all-or-nothing behaviour between them
```

Comment:
Actually no.. the execute (or task) will not add any more elements...


Still you are guaranteeing that STATE_NOT_RUNNING.



Also, I wanted to keep the shutdown returning a list of Threads, just like ExecutorService does... so callers could take different actions eventually.

Besides there shouldn't be many tasks pending by the time you called shutdown.

Reply:
I've found a very small case when it can happen but I'm not sure to be able to reproduce it...
About the API is only a suggestion because the original API of the executor was created prior of Java 8 and doesn't use functional interfaces to make things simpler/more efficient.
We can always provide this with ease:
```
default List<Runnable> shutdownNow(){
     final List<Runnable> tasks = new ArrayList<>();
     shutdownNow(tasks::add);
     return tasks;
}
```